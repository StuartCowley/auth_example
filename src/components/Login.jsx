import { useContext, useState } from "react"
import { useNavigate } from "react-router-dom"
import fakeLogin from "../utils/fakeLogin"
import Header from "./Header"
import jwtDecode from "jwt-decode"
import Cookie from "js-cookie"

import AuthContext from "../utils/AuthContext"

const Login = () => {
    const [details, setDetails] = useState({ username: "", email: "", password: "" })
    const [error, setError] = useState(null)
    const { setUser } = useContext(AuthContext)
    const navigate = useNavigate()

    const handleChange = ({ target: { value, id } }) => {
        setDetails(prev => ({ ...prev, [id]: value }))
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        /* Normally the password (in details) would be hashed with bcrypt
        This example uses hardcoded data without the backend but you must always
        remember to hash your passwords before sending them to the backend */
        const res = fakeLogin(details)
        if (res.error) {
            setError(res.error)
        } else {
            const currentUser = jwtDecode(res.token)
            console.log(currentUser, "<-- currentUser")
            setUser(currentUser)
            setError(null)
            // expiry date is expressed in days. 1/24 == 1 hour expiry time
            Cookie.set("token", res.token, {expires: 1/24})
            navigate('/account')
        }
    }

    return (
        <div>
            <Header />
            <p className="center">Please enter your details below to login</p>
            <div className="container">
                <form onSubmit={handleSubmit}>
                    <label htmlFor="username">Username</label><br />
                    <input id="username" onChange={handleChange} /><br />
                    <label htmlFor="email">Email</label><br />
                    <input id="email" onChange={handleChange} /><br />
                    <label htmlFor="password">Password</label><br />
                    <input id="password" type="password" onChange={handleChange} /><br />
                    <button type="submit">Login</button>
                    {error && <p className="error">{error}</p>}
                </form>
            </div>
        </div>
    )
}

export default Login