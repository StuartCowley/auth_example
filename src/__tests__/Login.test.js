import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import Login from "../components/Login";

describe("Login form component", () => {
    it("matches snapshot", () => {
        const generated = renderer
            .create(
                <Router>
                    <Login />
                </Router>
            ).toJSON();

        expect(generated).toMatchSnapshot();
    });

    it("title renders as expected", () => {
        render(
            <Router>
                <Login />
            </Router>
        );

        expect(screen.getByText("Please enter your details below to login")).toBeInTheDocument;
    });

    it("renders all form elements", () => {
        render(
            <Router>
                <Login />
            </Router>
        );
        const usernameField = screen.getByText("Username");
        const emailField = screen.getByText("Email");
        const passwordField = screen.getByText("Password");
        const loginButton = screen.getByRole("button", { name: "Login" });

        expect(usernameField).toBeInTheDocument;
        expect(emailField).toBeInTheDocument;
        expect(passwordField).toBeInTheDocument;
        expect(loginButton).toBeInTheDocument;
    });

    it("shows login failure message with non existent user", async () => {
        render(
            <Router>
                <Login />
            </Router>
        );

        const usernameField = screen.getByLabelText("Username");
        const emailField = screen.getByLabelText("Email");
        const passwordField = screen.getByLabelText("Password");
        const loginButton = screen.getByRole("button", { name: "Login" });

        fireEvent.change(usernameField, { target: { value: "test" } });
        fireEvent.change(emailField, { target: { value: "test@email.com" } });
        fireEvent.change(passwordField, { target: { value: "testPassword" } });
        fireEvent.click(loginButton);

        const errorMessage = await screen.findByText("details are incorrect");
        expect(errorMessage).toBeInTheDocument;
    });
});