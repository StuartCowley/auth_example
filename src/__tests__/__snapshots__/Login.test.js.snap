// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form component matches snapshot 1`] = `
<div>
  <nav
    className="header"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <div
      className="header__nav-link"
      onClick={[Function]}
    >
      Account
    </div>
    <a
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </nav>
  <p
    className="center"
  >
    Please enter your details below to login
  </p>
  <div
    className="container"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="username"
      >
        Username
      </label>
      <br />
      <input
        id="username"
        onChange={[Function]}
      />
      <br />
      <label
        htmlFor="email"
      >
        Email
      </label>
      <br />
      <input
        id="email"
        onChange={[Function]}
      />
      <br />
      <label
        htmlFor="password"
      >
        Password
      </label>
      <br />
      <input
        id="password"
        onChange={[Function]}
        type="password"
      />
      <br />
      <button
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
